File: app/src/main/java/de/eloc/eloc_control_panel/models/ElocDeviceInfo.java
package de.eloc.eloc_control_panel.models;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class ElocDeviceInfo {
    public final String plusCode;
    public final String name;
    public final double batteryVolts;
    public final String ranger;
    public final String time;
    public final double recTime;
    public final double accuracy;

    public ElocDeviceInfo(String plusCode, String name, double batteryVolts, String ranger, String time, double recTime, double accuracy) {
        this.plusCode = plusCode;
        this.name = name;
        this.batteryVolts = batteryVolts;
        this.ranger = ranger;
        this.time = time;
        this.recTime = recTime;
        this.accuracy = accuracy;
    }

    public static ArrayList<ElocDeviceInfo> parseForRanger(String json, String rangerName) {
        ArrayList<ElocDeviceInfo> deviceInfos = new ArrayList<>();

        try {
            JSONObject obj = new JSONObject(json);
            JSONArray deviceArray = obj.getJSONArray("device_infos");
            for (int i = 0; i < deviceArray.length(); i++) {
                JSONObject jsonInfo = deviceArray.getJSONObject(i);
                String ranger = jsonInfo.getString("ranger_name");
                if (!ranger.equals(rangerName)) {
                    // Only get devices for ranger with matched name.
                    continue;
                }

                String plusCode = jsonInfo.getString("plus_code");
                String deviceName = jsonInfo.getString("device_name");

                String time = jsonInfo.getString("time");
                double batteryVolts = 0.0;
                try {
                    String tmp = jsonInfo.getString("battery_volts");
                    batteryVolts = Double.parseDouble(tmp);
                } catch (NumberFormatException ignore) {

                }
                double recTime = 0.0;
                try {
                    String tmp = jsonInfo.getString("record_time_since_boot");
                    recTime = Double.parseDouble(tmp);
                } catch (NumberFormatException ignore) {

                }

                double accuracy = 0.0;
                try {
                    String tmp = jsonInfo.getString("gps_accuracy");
                    accuracy = Double.parseDouble(tmp);
                } catch (NumberFormatException ignore) {

                }

                ElocDeviceInfo deviceInfo = new ElocDeviceInfo(plusCode, deviceName, batteryVolts, ranger, time, recTime, accuracy);
                deviceInfos.add(deviceInfo);
            }
            Log.d("TAG", "parse: " + obj.getClass().getName());
        } catch (JSONException ignore) {
            ignore.printStackTrace();
        }

        return deviceInfos;
    }
}

File: app/src/main/java/de/eloc/eloc_control_panel/ng3/data/ElocDeviceInfo.kt
package de.eloc.eloc_control_panel.ng3.data

import org.json.JSONException
import org.json.JSONObject
import java.util.ArrayList

class ElocDeviceInfo(
    val plusCode: String,
    val name: String,
    val batteryVolts: Double,
    val ranger: String,
    val time: String,
    val recTime: Double,
    val accuracy: Double
) {
    companion object {

        fun parseForRanger(json: String, rangerName: String): ArrayList<ElocDeviceInfo> {
            val deviceInfos = ArrayList<ElocDeviceInfo>()

            try {
                val obj = JSONObject(json)
                val deviceArray = obj.getJSONArray("device_infos")
                for (i in 0 until deviceArray.length()) {
                    val jsonInfo = deviceArray.getJSONObject(i)
                    val ranger = jsonInfo.getString("ranger_name")
                    if (ranger != rangerName) {
                        // Only get devices for ranger with matched name.
                        continue
                    }

                    val plusCode = jsonInfo.getString("plus_code")
                    val deviceName = jsonInfo.getString("device_name")

                    val time = jsonInfo.getString("time")
                    var batteryVolts = 0.0
                    try {
                        val tmp = jsonInfo.getString("battery_volts")
                        batteryVolts = tmp.toDouble()
                    } catch (_: NumberFormatException) {

                    }
                    var recTime = 0.0
                    try {
                        val tmp = jsonInfo.getString("record_time_since_boot")
                        recTime = tmp.toDouble()
                    } catch (_: NumberFormatException) {

                    }

                    var accuracy = 0.0
                    try {
                        val tmp = jsonInfo.getString("gps_accuracy")
                        accuracy = tmp.toDouble()
                    } catch (ignore: NumberFormatException) {

                    }

                    val deviceInfo = ElocDeviceInfo(
                        plusCode,
                        deviceName,
                        batteryVolts,
                        ranger,
                        time,
                        recTime,
                        accuracy
                    )
                    deviceInfos.add(deviceInfo)
                }
            } catch (_: JSONException) {

            }

            return deviceInfos
        }
    }
}
